### PR TITLE
Improvements in the Associations Edit Workflow

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for New Issue (see also https://github.com/joomla/joomla-cms/pull/12308).
### Summary of Changes

In a multilingual install with associations, there are several issues with the item language and associations:
1. You cannot add associations when creating an item. Currently you need to: 
   - Create and save the item
   - Add the associations and save again
2. You cannot change a content language when editing an item. Currently you need to:
   - Edit the item and remove the associations first and save the item
   - Edit the item again, change the content language and save again
   - Finnaly, Edit the item again, change the associations and save again
3. You cannot add associations for unpublished content languages (but you have them across the backoffice)

Clearly, IMHO, this is not a very usable workflow and it comes with issues.

This PR should make that workflow a lot more easy, just add associations, and change the language whenever you need and save (one time).
Also allows to add associations for unpublished content languages.

Note: the only exception is the create/edit modals that will continue to not have Associations tab.
### Testing Instructions

1, Do a code review (hint: use https://github.com/joomla/joomla-cms/pull/12344/files?w=1 to ignore whitespace changes)
2. Use latest 3.7.x, multilingual with associations enabled
3. Apply patch 
4. For article, category, contact, menu item and newsfeed items check:
- When creating an item (try changing the languages, adding associations)
- When editing an item (try changing the languages, adding associations)
### Documentation Changes Required

None.
### Notes

Introduced two new methods: `JLanguageHelper::getContentLanguages` (get the content languages from `#__languages`) and `JLanguageHelper::getInstalledLanguages` (get the installed languages from `#__extensions` table).

Why this methods and not using `JLanguageHelper::getLanguages()`? 
- because `JLanguageHelper::getLanguages()` always return the published languages ... there are several cases we need the unpublisehd ones too ...
- JLanguageHelper::getLanguages() **does not check if the content languages are installed** which, of course, can, and will, lead to errors if you, for instance, uninstall a language package without removing the content language.
- because `JLanguageHelper::getLanguages()` is not IMHO totally semantic correct: get what languages? installed languages? content languages?
- among more things

This two methods will allow to do all languages related stuff across the core in the future (replacing several hardcoded queries, etc). I intend to do that step by step is this accepted.

@infograf768 as talked

### CODE REVIEW

File: administrator/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -260,7 +260,8 @@ JGLOBAL_ARTICLE_MANAGER_ORDER="Ordering"
 JGLOBAL_ARTICLE_ORDER_DESC="The order that articles will show in."
 JGLOBAL_ARTICLE_ORDER_LABEL="Article Order"
 JGLOBAL_ARTICLES="Articles"
-JGLOBAL_ASSOC_NOT_POSSIBLE="To define associations, please first save this item and make sure its language is not set to 'All'."
+JGLOBAL_ASSOC_NOT_POSSIBLE="To define associations, please make sure this item language is not set to 'All'."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: It should be either
This items language
Or
The item language
Or
The item's language

**Comment by andrepereiradasilva** [comment]
don't really know why drone is failing, something to do with `Highlight` js ... doesn't seem related to this PR

**Comment by zero-24** [comment]
Nope it is not your fault see: https://github.com/joomla/joomla-cms/pull/12345#issuecomment-252306596

**Comment by andrepereiradasilva** [comment]
now that PR is merged can a mantainer restart tarvis here?

**Comment by infograf768** [comment]
@zero-24 
can you relaunch Travis here please?

**Comment by brianteeman** [comment]
I restarted drone. No point in restarting travis as it passed

**Comment by zero-24** [comment]
It is a problem with the 3.7.x branch that needs syncing staging into 3.7.x this failed test has nothing to do with this PR.

**Comment by brianteeman** [comment]
I tried to restart Drone but it just says pending

**Comment by zero-24** [comment]
This can't be fixed with restarting drone as it is a problem in the 3.7.x branch ;) @brianteeman

**Comment by andrepereiradasilva** [comment]
ok. so ignore it, because it pass cs and unit tests.
can be tested

**Comment by infograf768** [comment]
will do asap. Your original branch was fine.

**Comment by infograf768** [comment]
I have not tested this item.<br /><br />Very clever implementation :+1: 

One more tester @alikon ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12344">issues.joomla.org/tracker/joomla-cms/12344</a>.</sub>

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on a59eb19a7fa67338ea6dc0aa86f618d234395264<br /><br />Very clever implementation :+1: 

One more tester @alikon ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12344">issues.joomla.org/tracker/joomla-cms/12344</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on a59eb19a7fa67338ea6dc0aa86f618d234395264<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12344">issues.joomla.org/tracker/joomla-cms/12344</a>.</sub>

**Comment by zero-24** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12344">issues.joomla.org/tracker/joomla-cms/12344</a>.</sub>

**Comment by infograf768** [comment]
do we have to test again?

**Comment by andrepereiradasilva** [comment]
don't think so, only fixes conflicts and updated branch

**Comment by infograf768** [comment]
I confirm.

@rdeutz
Please merge asap as we can't change an item language (when associated) without it.

**Pull request merged**